makebdb/makebdb.1
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\" makebdb.1                                                   May 2002
.\" ChrisS                                              chriss@pipex.net
.\"
.\" perdition
.\" Mail retrieval proxy server
.\" Copyright (C) 1999-2005  Horms <horms@verge.net.au>
.\" 
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of the
.\" License, or (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software Foundation,
.\" Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.TH MAKEBDB 1 "30th May 2002"
.SH NAME
makebdb \- Make a BerkeleyDB database from stdin 
.SH SYNOPSIS
\fBmakebdb\fP [options] bdbname
.SH DESCRIPTION
\fBmakebdb\fP allows a BerkeleyDB database to be formed
from input read from stdin. \fBbdbname\fP is the name of the BerkeleyDB
database to create, ususally the name of this file will end in .db. If the
file already exists is will be overwritten. The format of the input is
key, value pairs separated by a colon (:). 
.PP
Example:
.PP
this is a key:this is a value
.SH OPTIONS
.TP
.B \-h, \--help: 
Print usage information.
.TP
.B \-u, \--undo: 
Print content of database file, one entry a line.
.SH SEE ALSO
The full documentation for BerkeleyDB is available on the Sleepycat
Software web site (http://www.sleepycat.com)
.SH AUTHOR
Original GDBM code by Horms <horms@verge.net.au>.
Modified for BerkeleyDB by Chris Stratford <chriss@pipex.net>.


ChangeLog
Change Log

perdition
Mail retrieval proxy server
Copyright (C) 1999-2009  Horms <horms@verge.net.au>
----------------------------------------------------------------------

3rd September 2009
(Simon Horman)
 * 1.18-rc1
  - Partially revert "Build: reduce useless linking"
  - Debian: remove libjain from perdition.shlibs as it no longer exists
  - Build: reduce useless linking
  - Debian: Don't include full GPL in Debian copyright file
  - Debian: Remove ancient dpkg call
  - Fix syntax error in perdition.8 man page. .ni should be .fi
  - Debian: Update version to 1.18-rc1
  - Debian: Reinstate date for 1.17-8 changelog entry
  - Update FSF address
  - Debian: Don't ignore errors in helper scripts
  - Debian: Use dh_prep instead of dh_clean -k
  - Debian: Use ${binary:Version} instead of ${Source-Version}
  - Debian: Don't ignore distclean errors
  - Build: Use a conditional to control the installation of the ldap schema
  - Fix appending of domain name
  - Fix greeting
  - Update sockaddr and peeraddr handling for IPv6
  - Debian: Use debhelper compatibility level 7
  - Debian: Use standards version 3.8.2
  - Debian: ldaps is supported
  - Debian: ldap v3 is supported
  - Debian: close #513461 on the next release
  - Check alt subject if the common name doesn't match
  - Process all CN entries
  - Don't use X509_NAME_oneline() as it is deprecated
  - Null terminate the buffer resulting from the use of BIO_s_mem()
  - Add group to getgrname error message
  - Enhance getgrnam error message
  - Don't accept STARTTLS command in imap4 mode if it hasn't been enabled
  - Debian: Remove spurious build dependency on libtool
  - Debian: Update mysql build dependency to libmysqlclient-dev
  - Use suffix rules
  - Don't use := in Makefile.am, it isn't portable
  - Resolve libtoolize warnings
  - Move configure.in to configure.ac
  - PAM_RHOST support for perdition
  - Remove Makefile.in files from repository
  - MySQL fallback servers
  - Remove libjain
  - CDB database library corrections
  - Remove config.h.in and perdition/db/Makefile.in
  - Fix Berkeley typos
  - Debian: Add build dependency on libcdb-dev for recently added CDB support
  - CDB support in perdition
  - Debian: Bug #507414 is already pending
  - ldap_search_s() may return LDAP_SERVER_DOWN on success.
  - Allow the ldap db to return any one or more of: username, server and port
  - Removed tag B
  - update tags
  - Subject: Debian: LSB formatted dependency info in init.d script
  - Add connection rate limiting to TODO
  - Fix buffer overdlow in options logging code
  - Add custom login-disabled messages to TODO
  - Get rid of libjain, it does nothing
  - Clean up install-hook for schema
  - Don't use mkinstalldirs as it may not exist
  - Exit with exit status of configure if it fails
  - Date: Tue, 09 Oct 2007 19:03:21 -0300
  - Update change logs for CVE-2007-5740
  - Update debian/changelog for 1.17.1-1 release
  - Update debian/changelog for 1.17-8 release
  - Verify that tag read from end-users is valid
  - Make debian init script log something informative if perdition is already running
  - Don't dereference sockname in main() if it isn't intialised
  - Pass include '\0' in the length passed to yp_match() in the NIS module
  - Update TODO
  - Fix typo in MySQL arguments
  - Use ca_path and ca_file for incoming connections
  - Remove real email address from example
  - Debian packaging update: Release perdition 1.17-7
  - Debian packaging update
  - Debian Packaging Update
  - Fix documentation of map_library in sample configuration file
  - add multibind
  - Roll back non-debian portion of previous commit
  - Debian Packaging
  - Reverse regex REG_ICASE patch as the lower_case config option can be used
  - Debian Packaging Update: Uploaded to Debian.Org
  - Update MySQL build dependancy to libmysqlclient15-dev
  - Make regex matches case insensitive.
  - Include perdition_globals.h in distribtion
  - Don't build ldap documentation by default
  - Fix building of daemon map
  - tls_XXX_force now sets tls_XXX.
  - signdness fixes. macro cleanups. added log_password option
  - signdness fixes. macro cleanups. added log_password option
  - Fix includes
  - Allow the LDAP protocol version to be specified
  - Escape data fed into mysql queries. Probably needed for other
  - Add libtoolize to autogen.sh
  - The documentation for the mysql (and postgresql) pompaps
  - Diable use of ldap_initialise() as it seems broken on some
  - multiple ports
  - per-user restrictions
  - Rearange the deamon maps network code into a separate library
  - Update build dependancies from the stone age
  - Update copyright

----------------------------------------------------------------------

22nd June 2005
(Horms)
 * 1.17
 - Added explicit domain option. Brendan O'Dea
 - Make sure all source files are inculded in tar ball,
   regardless of what parts of the code are excluded from the build
   by configure.
 - Fixed possible buffer overflow in berkely db module
   Chris Stratford
 - Option to disable compilation of ldap (latex) documentation
 - Use ldap_initialize() when compiled against openssl (>=2.0.0)
   to allow arbitary ldap urls. In particular, ldaps is now supported.
 - Don't add libraries discovered in configure to LDADD.
 - Better loging connection failures relating to capabilities
   on the real-server
 - Pass asynchronous messages sent by the real-server during
   authentication back to the end-user
 - When add_domain is used, allow the depth to which the domain
   name has leading levels stripped.
 - Fix gcc-4.0 compile problem. Andreas Jochens
 - Removed bogus loging that occurs before the logger is
   fully intialised when debug is enabled. Roberto Suarez Soto
 - Add link daemon module against socket libraries
 - Reopen logger so child processes get their own file descriptors
 - Fix segmentation fault when timeout is set to 0. John Capo
 - Fix ODBC library detection
 - Fix printing of server_port ojects (in particular the logging

makebdb/makebdb.c
/**********************************************************************
 * makebdb.c                                                   May 2002
 * ChrisS                                              chriss@pipex.net
 *
 * Create a bdb file from a : delimited flat file read from stdin
 * Based (very heavily) on makegdbm.c by Horms
 *
 * perdition
 * Mail retrieval proxy server
 * Copyright (C) 1999-2005  Horms
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 *
 **********************************************************************/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <ctype.h>
#include <db.h>

#include "options.h"

#ifdef DMALLOC
#include <dmalloc.h>
#endif


#define MAX_LINE_LENGTH 4096
#define FIELD_DELIMITER ':'

void makebdb_new(const makebdb_options_t options);
void makebdb_undo(const makebdb_options_t options);

int main(int argc, char **argv){
  makebdb_options_t options;

  options=makebdb_options(argc, argv);


  if(options.undo){
    makebdb_undo(options);
  }
  else{
    makebdb_new(options);
  }

  return(0);
}
  

/**********************************************************************
 * makebdb_new
 * Create a new bdb file from a flat file read from stdin
 * pre: options: options structure specifying what to do
 * post: bdb file created and populated
 **********************************************************************/

void makebdb_new(const makebdb_options_t options){
  DB *dbp;
  DBT key;
  DBT content;
  int ret;
  int i;
  char line[MAX_LINE_LENGTH];
  int status;
  int blank;
  int lineno;

  if ((ret=db_create(&dbp, NULL, 0)) != 0) {
    fprintf(stderr,"makebdb_new: db_create: %s\n",db_strerror(ret));
    exit(-1);
  }
  if ((ret=dbp->open(
    dbp,
#ifdef HAVE_BDB_4_1
    NULL,
#endif
    options.mapname,
    NULL,
    DB_HASH,
    DB_CREATE|DB_TRUNCATE,
    0644
  )) != 0) {
    fprintf(stderr,"makebdb_new: DB->open: %s\n",db_strerror(ret));
    exit(-1);
  }

  /*Warning: This loop contains some of the worst code ever written*/
  memset(&key, 0, sizeof(key));
  memset(&content, 0, sizeof(content));
  status=0;
  lineno=0;
  while(fgets(line, MAX_LINE_LENGTH, stdin)!=NULL){
    blank=1;
    key.data=line;
    key.size=-1;
    content.size=-1;
    for(i=0;i<MAX_LINE_LENGTH;i++){
      if(blank && *(line+i)!=' ' && *(line+i)!='\t'&& *(line+i)!='\n'){
        blank=0;
      }
      if(key.size==-1 && *(line+i)==FIELD_DELIMITER){
        key.size=i;
        content.data=line+i+1;
        continue;
      }
      if(content.size==-1 && *(line+i)=='\n'){
        content.size=i-key.size-1;
        lineno++;
        break;
      }
    }
    fflush(NULL);
    if(!blank && !status && (key.size<1 || content.size<1)){
      status=1;
    }
    if(status){
      if(*(line+i)=='\n'){
        fprintf(stderr, "makebdb: invalid input on line: %d \n", lineno);
        status=0;
      }
      continue;
    }

    if(blank){
      continue;
    }
    if ((ret = dbp->put(dbp, NULL, &key, &content, DB_NOOVERWRITE))!=0){
      *((char *)key.data+key.size)='\0';
      fprintf(stderr, "makebdb: %s (key: %s)\n",db_strerror(ret),
		      (char *)key.data);
    }

  }

  dbp->close(dbp, 0);
}


/**********************************************************************
 * makebdb_undo
 * print out a : delimited flat file to stdout of a bdb file
 * pre: options: options structure sepcifying what to do
 * post: glat file output to stdout
 **********************************************************************/

void makebdb_undo(makebdb_options_t options){
  DB *dbp;
  DBC *dbcp;
  DBT key;
  DBT content;
  int ret;
  
  if ((ret=db_create(&dbp, NULL, 0)) != 0) {
    fprintf(stderr,"makebdb_undo: db_create: %s\n",db_strerror(ret));
    exit(-1);
  }
  if ((ret=dbp->open(
    dbp,
#ifdef HAVE_BDB_4_1
    NULL,
#endif
    options.mapname,
    NULL,
    DB_HASH,
    DB_RDONLY,
    0644
  )) != 0) {
    fprintf(stderr,"makebdb_undo: DB->open: %s\n",db_strerror(ret));
    exit(-1);
  }

  if ((ret = dbp->cursor(dbp, NULL, &dbcp, 0))!=0) {
    fprintf(stderr,"makebdb_undo: DB->cursor: %s\n",db_strerror(ret));
    exit(-1);
  };

  memset(&key, 0, sizeof(key));
  memset(&content, 0, sizeof(content));

  while((ret=dbcp->c_get(dbcp,&key,&content,DB_NEXT))==0)
  {
    char c;
    fwrite((char *)key.data,  key.size, 1, stdout);
    c = ':';
    fwrite(&c, 1, 1, stdout);
    fwrite((char *)content.data,  content.size, 1, stdout);
    c = '\n';
    fwrite(&c, 1, 1, stdout);
  }
  if(ret != DB_NOTFOUND){
    fprintf(stderr,"makebdb_undo: DBcursor->get: %s\n",db_strerror(ret));
  }

  dbcp->c_close(dbcp);
  dbp->close(dbp, 0);
} 
